
# PRACTICA NLP. NOTEBOOK 4

## 4. Reporte de métricas y conclusiones

El alumno, tomando como referencia los resultados del modelo escogido en el ejercicio 3, calculará las métricas que permitan validar la bondad del modelo. 

También incluirá comentarios y las conclusiones finales.

Del notebook anterior concluyo que, aunque mejorable, el mejor modelo que he probado es una clasificación random forest, con el preprocesado definido en el notebook 2 y los valores de algunos parámetros del tfidf ajustados.

La matriz de confusión en la que se muestran los resultados de aciertos y fracasos en la predicción es la siguiente: 

In [1]:
import pandas as pd
conf_matrix= pd.read_csv('confusion_matrix.csv', sep=',', decimal='.')
conf_matrix

overall  0.0  1.0
0      0.0  804  184
1      1.0  276  736

In [2]:
#retoco los nombres de las columnas y filas de la matriz para facilitar la comprensión
conf_matrix.rename(columns={'overall': 'label', '0.0': 'negative','1.0':'positive'}, inplace=True)
conf_matrix.set_axis(['negative','positive'], axis='index')


label  negative  positive
negative    0.0       804       184
positive    1.0       276       736

In [3]:
#hago lo mismo con el classification report
classification_report= pd.read_csv('classification_report.csv', sep=',', decimal='.')
classification_report.rename(columns={'Unnamed: 0': 'label & global_metrics'}, inplace=True)
classification_report.set_axis(['negative','positive','accuracy','promedio','media ponderada'], axis='index')

label & global_metrics  precision    recall  f1-score  support
negative                           0.0   0.744444  0.813765  0.777563   988.00
positive                           1.0   0.800000  0.727273  0.761905  1012.00
accuracy                      accuracy   0.770000  0.770000  0.770000     0.77
promedio                     macro avg   0.772222  0.770519  0.769734  2000.00
media ponderada           weighted avg   0.772556  0.770000  0.769640  2000.00

Se seleccionaron para entrenar el modelo 10000 entradas del dataset de reviews de instrumentos musicales, 5000 con valoración positiva y 5000 con valoración negativa, para poder entrenar un modelo balanceado.

De estas valoraciones 2000 se reservaron para la fase de test y 8000 se utilizaron para el entrenamiento, como vimos en el notebook anterior. 

En la matriz de confusión y en el classification report podemos observar que el conjunto de test tiene 988 valoraciones negativas y 1012 positivas. 

Después del entrenamiento los datos globales son los que vemos en la matriz de confusión: 

-el modelo interpreta correctamente como negativas 804 de las 988 y se equivoca en 184 que ve positivas cuando no lo son y

-el modelo interpreta correctamente como positivas 736 y se equivoca en 276 que ve como negativas cuando no lo son. 

El classification report muestra las métricas de precision, recall y f1-score por clase (en las dos primeras filas, 0-negativa, 1-positiva, puesto que se trata de una clasificación binaria).
Las métricas se calculan utilizando los valores de la matriz de confusión, true y false positives, y true y false negatives. 
Hay 4 maneras de comprobar si las predicciones son o no correctas:

TN / True Negative: review negativa predicha como negativa
TP / True Positive: review positiva y predicha positiva
FN / False Negative: review positiva pero predicha negativa
FP / False Positive: review negativa pero predicha positiva


Precision: ¿qué porcentaje de las predicciones son correctas? 
Recall: qué porcentaje de las review positivas detecta el modelo? 
F1 score: qué porcentaje de predicciones positivas eran correctas? 

Precision = TP/(TP + FP)
Recall = TP/(TP+FN)
F1 Score = 2*(Recall * Precision) / (Recall + Precision)

##### Resumen de resultados: 

El modelo identifica correctamente el sentimiento de una review el 77% del tiempo. ->accuracy
Cuando ha predicho una review como positiva, esa review lo era realmente el 80% de las veces -> precision
Cuando una review era positiva, el modelo lo ha identificado correctamente en el 72.7% de los casos.->recall

macro-avg es el promedio de precision/recall/F1 de todas las clases y weighted avg la media ponderada del total de TP/total de reviews: ambas 0.77 en este caso.

Aunque era la mejor versión conseguida hasta el momento en cuanto a resultados comparado con otros modelos, se puede mejorar mucho. 


 ### Conclusiones Globales y Mejoras posibles:
  
 - El resumen de resultados del punto anterior es una buena conclusión. Teniendo en cuenta que detecto muchas cosas aún por hacer, más que concluir me centro en las mejoras. 
 
 - En primer lugar, empezando por el final, en los modelos se podía haber jugado con los valores de los parámetros, además de probar diferentes modelos y modificar los valores del tfidf en bloque (versión default/customizada)
 
 - También tenía localizado un modelo preentrenado de redes neuronales que quería probar.. para otra ocasión.
 
 - Por otro lado, se ha filtrado el número de reviews que se han utilizado para poder obtener resultados de una manera más rápida en los cálculos, pero seguro que con más tiempo se puede hacer con una mayor cantidad de datos en la muestra a pesar de las limitaciones del pc. Otra cosa que se podría haber probado ahora al final con el modelo, es introducirle nuevos datos de los que han quedado fuera y ver qué tal generaliza.
 
 - En cuanto al dataset, decidí en un primer momento juntar el summary con el texto de la review y quizá no ha ayudado al resultado al realizar el filtro. De hecho, una idea de mejora es equilibrar el número de caracteres de las reviews. Me queda pendiente hacer pruebas limitando el número de caracteres, creo que hay reviews muy largas difíciles de interpretar. 
 
 - Como ya comentaba en el notebook del preprocesado, encontré un listado de stopwords  muy completo pero de un estudio realizado sobre el nytimes. de haber tenido más tiempo habría localizado un listado de stopwords más informal. Finalmente para una de las versiones trabajé con el listado descargado y para otra con el listado de spacy. Esta última es la que ha quedado para el notebook 3. Me hubiera gustado probar con el mío además de buscar otro.
 
 - En el primer notebook planteé la opción de revisar si influyen las personas en el tipo de review (si los que escriben muchas reviews tienden a poner la mayoría positivas o negativas...a alguna marca en particular..tipos de personas, de influencias..)...y si hay algún efecto observable en las reviews. Para prácticas futuras. 
 
 - Por último, como también comentaba en un principio, encontré dos listados de palabras uno formado por palabras positivas y otro formado por palabras negativas que me hubiera gustado utilizar para analizar las valoraciones iniciales con un 3, para decidir si las asociaba a la valoración negativa o positiva con un fundamento más allá de mi propio criterio. 
  
 ... Tampoco he podido probar con los otros dos datasets descargados, para valorar la capacidad de generalización del modelo. 
 
 Muchas cosas por mejorar, pero muy contenta de haber llegado hasta aquí. 
 Gracias!
 
 